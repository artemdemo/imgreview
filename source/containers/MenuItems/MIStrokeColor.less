.main-menu-color {
    display: inline-block;
    width: 20px;
    height: 18px;
    vertical-align: bottom;
}
