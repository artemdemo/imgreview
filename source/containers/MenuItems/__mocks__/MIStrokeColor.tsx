import React from 'react';

const MIStroke = props => (
    <div data-mock='MIStroke'>
        {props.children}
    </div>
);

export default MIStroke;
