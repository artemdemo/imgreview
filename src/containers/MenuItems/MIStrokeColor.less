.MIStrokeColor__Current {
  display: inline-block;
  width: 20px;
  height: 18px;
  vertical-align: bottom;
}
