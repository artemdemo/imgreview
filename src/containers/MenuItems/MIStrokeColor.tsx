import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import _ from 'lodash';
import * as canvasApi from '../../../srcCanvas/api';
import { TReduxState } from '../../reducers';
import { TStateMenu } from '../../model/menu/menuReducer';
import TopMenuItem from '../../components/TopMenu/TopMenuItem';
import ColorSelector from '../ColorSelector/ColorSelector.async';
import { showColorPicker, setStrokeColor } from '../../model/menu/menuActions';
import store from '../../store';
import * as gaService from '../../services/ganalytics';

const getShapeColor = (shape: any) => {
  if (_.isFunction(shape.getStrokeColor)) {
    return shape.getStrokeColor();
  }
  if (_.isFunction(shape.getFillColor)) {
    return shape.getFillColor();
  }
  throw new Error("Can't get shape color");
};

canvasApi.shapeClicked.on((shape: canvasApi.TShapeClicked) => {
  const shapeColor = getShapeColor(shape);
  store.dispatch(setStrokeColor(shapeColor));
});

const MIStrokeColor__Current = styled.div`
  display: inline-block;
  width: 20px;
  height: 18px;
  vertical-align: bottom;
`;

type TProps = {
  menu: TStateMenu;
  showColorPicker: () => void;
  disabled: boolean;
  show: boolean;
};

type TState = {
  loading: boolean;
};

class MIStrokeColor extends React.PureComponent<TProps, TState> {
  static readonly defaultProps = {
    disabled: false,
    show: false,
  };

  state = {
    loading: true,
  };

  onClick = () => {
    const { showColorPicker } = this.props;
    showColorPicker();
  };

  colorSelectorIsReady = () => {
    this.setState({
      loading: false,
    });
  };

  handleChangeStrokeColor = (color: string) => {
    gaService.sendEvent({
      eventCategory: gaService.EEventCategories.MenuClick,
      eventAction: gaService.EEventActions.ChangeColor,
    });
  };

  renderColorSelector() {
    return (
      <ColorSelector
        isLoaded={this.colorSelectorIsReady}
        onChangeStrokeColor={this.handleChangeStrokeColor}
      />
    );
  }

  render() {
    const { disabled, show, menu } = this.props;
    if (!show && this.state.loading) {
      // Here I'm rendering color selector in order to kick-in lazy loading.
      // Color selector wouldn't be shown.
      // After ColorSelector is loaded I can stop rendering it (hence, the `if` statement)
      return this.renderColorSelector();
    }
    return (
      <TopMenuItem
        onClick={this.onClick}
        show={show}
        disabled={disabled || this.state.loading}
      >
        <MIStrokeColor__Current
          style={{
            backgroundColor: menu.strokeColor,
          }}
        />
        {this.renderColorSelector()}
      </TopMenuItem>
    );
  }
}

export default connect(
  (state: TReduxState) => ({
    menu: state.menu,
  }),
  {
    showColorPicker,
  }
)(MIStrokeColor);
