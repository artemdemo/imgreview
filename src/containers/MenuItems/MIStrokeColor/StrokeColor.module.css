.StrokeColor {
  display: flex;
  width: 19px;
  height: 18px;
  vertical-align: bottom;
  border-radius: 2px;
}

.StrokeColor_isTransparent {
  border: 1px solid rgba(0, 0, 0, 0.15);
}

.StrokeColor__Plug {
  background-color: var(--main-menu-color);
  width: 65%;
  height: 65%;
  margin: auto;
  border-radius: 1px;
}

.StrokeColor__Plug_isTransparent {
  border: 1px solid rgba(0, 0, 0, 0.15);
}
