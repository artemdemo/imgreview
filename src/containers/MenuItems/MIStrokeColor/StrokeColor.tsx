import React, { useContext, useEffect, useState } from 'react';
import Color from 'color';
import classnames from 'classnames';
import { AppStateContext } from '../../../model/AppStateContext';
import s from './StrokeColor.module.css';

export const StrokeColor: React.FC = () => {
  const {
    state: {
      menu: { strokeColor },
    },
  } = useContext(AppStateContext);
  const [isTransparent, setIsTransparent] = useState(false);

  useEffect(() => {
    setIsTransparent(Color(strokeColor).alpha() < 0.15);
  }, [strokeColor]);

  return (
    <div
      className={classnames({
        [s.StrokeColor]: true,
        [s.StrokeColor_isTransparent]: isTransparent,
      })}
      style={{
        backgroundColor: strokeColor,
      }}
    >
      <div
        className={classnames({
          [s.StrokeColor__Plug]: true,
          [s.StrokeColor__Plug_isTransparent]: isTransparent,
        })}
      />
    </div>
  );
};
