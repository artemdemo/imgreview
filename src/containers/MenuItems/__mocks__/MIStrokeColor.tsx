import React from 'react';

const MIStrokeColor = (props) => (
  <div data-mock="MIStrokeColor">{JSON.stringify(props, null, 2)}</div>
);

export default MIStrokeColor;
