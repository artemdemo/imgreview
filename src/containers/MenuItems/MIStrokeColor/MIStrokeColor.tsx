import React, { useContext, useState } from 'react';
import _ from 'lodash';
import dynamic from 'next/dynamic';
import { TopMenuItem } from '../../../components/TopMenu/TopMenuItem';
import { setStrokeColor } from '../../../model/menu/menuActions';
import * as gaService from '../../../services/ganalytics';
import { t } from '../../../services/i18n';
import IShape from '../../../../srcCanvas/canvasShapes/Shape/IShape';
import { AppStateContext } from '../../../model/AppStateContext';
import { StrokeColor } from './StrokeColor';
import { useShapeTouched } from '../../../hooks/useShapeTouched';

const ColorSelector = dynamic(
  () =>
    import(
      /* webpackChunkName: "ColorSelector" */
      '../../ColorSelector/ColorSelector'
    ),
  { loading: () => null, ssr: false },
);

const getShapeStrokeColor = (shape: IShape): string | undefined => {
  if (_.isFunction(shape.getStrokeColor)) {
    return shape.getStrokeColor();
  }
};

type Props = {
  disabled?: boolean;
};

export const MIStrokeColor: React.FC<Props> = (props) => {
  const { disabled = false } = props;
  const {
    state: {
      menu: { strokeColor },
      canvas: { canvasApi },
    },
    dispatch,
  } = useContext(AppStateContext);
  const [showColorPicker, setShowColorPicker] = useState(false);

  useShapeTouched((shape: IShape) => {
    const shapeColor = getShapeStrokeColor(shape);
    if (shapeColor) {
      dispatch(setStrokeColor(shapeColor));
    }
  });

  return (
    <TopMenuItem
      onClick={() => {
        setShowColorPicker(true);
      }}
      title={t('menu.strokeColor')}
      disabled={disabled}
      satellite={
        <ColorSelector
          show={showColorPicker}
          onHide={() => {
            setShowColorPicker(false);
          }}
          color={strokeColor}
          onChange={(color: string) => {
            dispatch(setStrokeColor(color));
            canvasApi?.setStrokeColorToActiveShape(color);

            gaService.sendEvent({
              eventCategory: gaService.EEventCategories.MenuClick,
              eventAction: gaService.EEventActions.ChangeStrokeColor,
              doNotRepeat: true,
            });
          }}
        />
      }
    >
      <StrokeColor />
    </TopMenuItem>
  );
};
