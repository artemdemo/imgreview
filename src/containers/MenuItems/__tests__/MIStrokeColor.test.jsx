import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import MIStrokeColor from '../MIStrokeColor';

jest.mock('react-redux');
jest.mock('../../../components/TopMenu/TopMenuItem');
jest.mock('../../ColorSelector/ColorSelector.async');

describe('MIStrokeColor', () => {
    jest.clearAllMocks();
    const reactReduxMock = require('react-redux');

    it('should render', () => {
        const state = {
            menu: {
                strokeColor: 'red',
            },
        };
        const tree = renderer.create(
            <MIStrokeColor
                menu={state.menu}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
        expect(reactReduxMock.__getLastMaps().mapStateToProps(state)).toEqual({
            menu: state.menu,
        });
    });

    it('should be disabled', () => {
        const tree = renderer.create(
            <MIStrokeColor
                menu={{
                    strokeColor: 'red',
                }}
                disabled
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should handle click', () => {
        const showColorPickerMock = jest.fn();
        const wrapper = mount(
            <MIStrokeColor
                menu={{
                    strokeColor: 'red',
                }}
                showColorPicker={showColorPickerMock}
            />
        );
        wrapper.find('[data-mock="TopMenuItem"]').simulate('click');
        expect(showColorPickerMock).toBeCalled();
    });
});
