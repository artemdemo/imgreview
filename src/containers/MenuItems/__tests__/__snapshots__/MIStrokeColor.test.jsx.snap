// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MIStrokeColor should be disabled 1`] = `
Array [
  <div
    data-mock="TopMenuItem"
    onClick={[Function]}
  >
    <div
      className="sc-AxjAm blVvcT"
      style={
        Object {
          "backgroundColor": "red",
        }
      }
    />
  </div>,
  <div
    data-mock="ColorSelector"
  />,
]
`;

exports[`MIStrokeColor should render 1`] = `
Array [
  <div
    data-mock="TopMenuItem"
    onClick={[Function]}
  >
    <div
      className="sc-AxjAm blVvcT"
      style={
        Object {
          "backgroundColor": "red",
        }
      }
    />
  </div>,
  <div
    data-mock="ColorSelector"
  />,
]
`;
